pkg/k8s/client_test.go

```diff
@@ -32,79 +45,166 @@ func TestGetSecret(t *testing.T) {
+			t.podList = append(t.podList[:i], t.podList[i+1:]...)
+			break
+		}
+	}
+}
 
-	podNs, podName, err := kubeClient.GetPodName(podIP)
+func (t *TestClientSet) SerializeObject(o interface{}) io.ReadCloser {
+	output, err := json.MarshalIndent(o, "", "")
+	if err != nil {
+		panic(err)
+	}
+	return ioutil.NopCloser(bytes.NewReader([]byte(output)))
+}
+
+func (t *TestClientSet) GetPodList() io.ReadCloser {
+
+	podList := &v1.PodList{}
+
+	for _, p := range t.podList {
+		podList.Items = append(podList.Items, p)
+	}
+
+	podList.TypeMeta = metav1.TypeMeta{
+		Kind:       "PodList",
+		APIVersion: "v1",
+	}
+
+	return t.SerializeObject(podList)
+}
+
+func TestGetPodInfo(t *testing.T) {
+
+	testClientSet := &TestClientSet{}
+	client, restClient := testClientSet.GetTestClientSet()
+
+	optionsModifier := func(options *metav1.ListOptions) {}
+	podListWatch := cache.NewFilteredListWatchFromClient(
+		restClient,
+		"pods",
+		v1.NamespaceAll,
+		optionsModifier,
+	)
+	kubeClient := &KubeClient{ClientSet: client, PodListWatch: podListWatch}
+
+	// Test a single pod
+	testPodName := "testpodname"
+	testPodNs := "default"
+	testPodIP := "10.0.0.8"
+	testClientSet.AddPod(testPodName, testPodNs, testPodIP)
+	podNs, podName, _, _, err := kubeClient.GetPodInfo(testPodIP)
 	if err != nil {
 		t.Fatalf("Error getting pod: %v", err)
 	}
-	if podName != "testpodname" {
+	if podName != testPodName {
 		t.Fatalf("Incorrect pod name: %v", podName)
 	}
-	if podNs != "default" {
+	if podNs != testPodNs {
 		t.Fatalf("Incorrect pod ns: %v", podNs)
 	}
+
+	// Delete test
+	testPodIP = "10.0.0.8"
+	testClientSet.DeletePod(testPodName, testPodNs)
+	podNs, podName, _, _, err = kubeClient.GetPodInfo(testPodIP)
+	if err == nil {
+		t.Fatal("Pod still in pod list")
+	}
 }
 
 func TestPodListRetries(t *testing.T) {
 	// this test is to solely test the retry and sleep logic works as expected
 	podIP := "10.0.0.8"
+	testClientSet := &TestClientSet{}
+	client, restClient := testClientSet.GetTestClientSet()
 
-	fakeClient := fake.NewSimpleClientset()
+	testPodName := "testpodname"
+	testPodNs := "default"
+	testPodIP := "10.0.0.8"
 
-	pod := &v1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "testpodname1",
-			Namespace: "default",
-		},
-		Status: v1.PodStatus{
-			PodIP: podIP,
-		},
-	}
-	kubeClient := &KubeClient{ClientSet: fakeClient}
+	optionsModifier := func(options *metav1.ListOptions) {}
+	podListWatch := cache.NewFilteredListWatchFromClient(
+		restClient,
+		"pods",
+		v1.NamespaceAll,
+		optionsModifier,
+	)
+
+	kubeClient := &KubeClient{ClientSet: client, PodListWatch: podListWatch}
 
 	time.AfterFunc(time.Duration(1200*time.Millisecond), func() {
-		fakeClient.CoreV1().Pods("default").Create(pod)
+		testClientSet.AddPod(testPodName, testPodNs, testPodIP)
 	})
 
 	start := time.Now()
-	podNs, podName, err := kubeClient.GetPodName(podIP)
+	podNs, podName, _, _, err := kubeClient.GetPodInfo(podIP)
```

Comment:
Should we also check the replica set name and labels are returned correctly?

Reply:
Will add in future PR.